File: features/show-details/ui/src/main/kotlin/com/thomaskioko/tvmaniac/showdetails/ui/section/ShowDetailsSeasonEpisodesSection.kt
package com.thomaskioko.tvmaniac.showdetails.ui.section

import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.tooling.preview.PreviewWrapper
import com.thomaskioko.tvmaniac.compose.components.ThemePreviews
import com.thomaskioko.tvmaniac.compose.components.TvManiacPreviewWrapperProvider
import com.thomaskioko.tvmaniac.presenter.showdetails.seasonsepisodes.ShowDetailsMarkEpisodeUnwatched
import com.thomaskioko.tvmaniac.presenter.showdetails.seasonsepisodes.ShowDetailsMarkEpisodeWatched
import com.thomaskioko.tvmaniac.presenter.showdetails.seasonsepisodes.ShowDetailsSeasonClicked
import com.thomaskioko.tvmaniac.presenter.showdetails.seasonsepisodes.ShowDetailsSeasonsEpisodesAction
import com.thomaskioko.tvmaniac.presenter.showdetails.seasonsepisodes.ShowDetailsSeasonsEpisodesPresenter
import com.thomaskioko.tvmaniac.presenter.showdetails.seasonsepisodes.ShowDetailsSeasonsEpisodesState
import com.thomaskioko.tvmaniac.showdetails.nav.model.ShowSeasonDetailsParam
import com.thomaskioko.tvmaniac.showdetails.ui.components.ContinueTrackingSection
import com.thomaskioko.tvmaniac.showdetails.ui.components.WatchProgressSection
import com.thomaskioko.tvmaniac.showdetails.ui.previewSeasonsEpisodesState
import com.thomaskioko.tvmaniac.testtags.showdetails.ShowDetailsTestTags

@Composable
internal fun ShowDetailsSeasonEpisodesSection(
    presenter: ShowDetailsSeasonsEpisodesPresenter,
    status: String?,
) {
    val state by presenter.state.collectAsState()
    ShowDetailsSeasonEpisodesSection(
        state = state,
        status = status,
        onAction = presenter::dispatch,
    )
}

@Composable
internal fun ShowDetailsSeasonEpisodesSection(
    state: ShowDetailsSeasonsEpisodesState,
    status: String?,
    onAction: (ShowDetailsSeasonsEpisodesAction) -> Unit,
) {
    ContinueTrackingSection(
        modifier = Modifier.testTag(ShowDetailsTestTags.CONTINUE_TRACKING_SECTION_TEST_TAG),
        episodes = state.continueTrackingEpisodes,
        scrollIndex = state.continueTrackingScrollIndex,
        updatingEpisodeIds = state.updatingEpisodeIds,
        onMarkWatched = { episode ->
            if (episode.isWatched) {
                onAction(
                    ShowDetailsMarkEpisodeUnwatched(
                        showId = episode.showId,
                        episodeId = episode.episodeId,
                    ),
                )
            } else {
                onAction(
                    ShowDetailsMarkEpisodeWatched(
                        showId = episode.showId,
                        episodeId = episode.episodeId,
                        seasonNumber = episode.seasonNumber,
                        episodeNumber = episode.episodeNumber,
                    ),
                )
            }
        },
    )
    WatchProgressSection(
        modifier = Modifier.testTag(ShowDetailsTestTags.WATCH_PROGRESS_SECTION_TEST_TAG),
        status = status,
        watchedEpisodesCount = state.watchedEpisodesCount,
        totalEpisodesCount = state.totalEpisodesCount,
        seasonsList = state.seasonsList,
        selectedSeasonIndex = state.selectedSeasonIndex,
        showHeader = state.continueTrackingEpisodes.isEmpty(),
        onSeasonClicked = { index, season ->
            onAction(
                ShowDetailsSeasonClicked(
                    params = ShowSeasonDetailsParam(
                        showId = season.tvShowId,
                        seasonId = season.seasonId,
                        seasonNumber = season.seasonNumber,
                        selectedSeasonIndex = index,
                    ),
                ),
            )
        },
    )
}

@ThemePreviews
@PreviewWrapper(TvManiacPreviewWrapperProvider::class)
@Composable
private fun ShowDetailsSeasonEpisodesSectionPreview() {
    ShowDetailsSeasonEpisodesSection(
        state = previewSeasonsEpisodesState,
        status = "Returning Series",
        onAction = {},
    )
}

File: ios/Packages/ShowDetails/Sources/ShowDetails/Sections/ShowDetailsSeasonEpisodesSection.swift
import Components
import DesignSystem
import Models
import SwiftUI
import TvManiacKit
import UpNext

struct ShowDetailsSeasonEpisodesSection: View {
    private let presenter: ShowDetailsSeasonsEpisodesPresenter
    private let showStatus: String?
    @StateValue private var state: ShowDetailsSeasonsEpisodesState

    init(presenter: ShowDetailsSeasonsEpisodesPresenter, showStatus: String?) {
        self.presenter = presenter
        self.showStatus = showStatus
        _state = .init(presenter.stateValue)
    }

    var body: some View {
        let showHeader = state.continueTrackingEpisodes.count == 0
        ContinueTrackingSection(
            title: String(\.title_continue_tracking),
            episodes: Array(state.continueTrackingEpisodes).map { $0.toSwift() },
            scrollIndex: Int(state.continueTrackingScrollIndex),
            dayLabelFormat: { count in String(\.day_label, quantity: count) },
            tbdLabel: String(\.label_tbd),
            onMarkWatched: { episode in
                if episode.isWatched {
                    presenter.dispatch(action: ShowDetailsMarkEpisodeUnwatched(
                        showId: episode.showId,
                        episodeId: episode.episodeId
                    ))
                } else {
                    presenter.dispatch(action: ShowDetailsMarkEpisodeWatched(
                        showId: episode.showId,
                        episodeId: episode.episodeId,
                        seasonNumber: episode.seasonNumber,
                        episodeNumber: episode.episodeNumber
                    ))
                }
            },
            updatingEpisodeIds: Set(Array(state.updatingEpisodeIds).map(\.int64Value))
        )
        SeasonProgressSection(
            title: String(\.title_season_details),
            showHeader: showHeader,
            status: showStatus,
            watchedEpisodesCount: Int32(state.watchedEpisodesCount),
            totalEpisodesCount: Int32(state.totalEpisodesCount),
            seasonsList: Array(state.seasonsList).map { $0.toSwift() },
            selectedSeasonIndex: Int(state.selectedSeasonIndex),
            seasonCountFormat: { count in String(\.season_count, quantity: Int(count)) },
            episodesWatchedFormat: { watched, total in
                String(\.episodes_watched, quantity: Int(total), Int(watched), Int(total))
            },
            episodesLeftFormat: { count in
                String(\.episodes_left, quantity: Int(count), Int(count))
            },
            upToDateLabel: String(\.label_up_to_date),
            onSeasonClicked: { index, season in
                let params = ShowSeasonDetailsParam(
                    showId: season.tvShowId,
                    seasonId: season.seasonId,
                    seasonNumber: season.seasonNumber,
                    selectedSeasonIndex: Int32(index)
                )
                presenter.dispatch(action: ShowDetailsSeasonClicked(params: params))
            }
        )
    }
}
